native/native_main.py

```diff
@@ -182,224 +660,387 @@ def win_firefox_restart(message):
+                "cmd": "error",
+                "error": "%s %s"
+                % (
+                    "Error decoding JSON object.",
+                    "Invalid 'prefs' list specified.",
+                ),
+            }
 
-        #
-        # subprocess.Popen(
-        #    [ff_bin_path, "-profile", profile_dir],
-        #    shell=False,
-        #    creationflags=0x208 \
-        #    | subprocess.CREATE_NEW_PROCESS_GROUP)
-        #
+        util = Utility()
+        prefs_to_remove = util.cleanup_items(prefs_to_remove)
 
-        #
-        # 'schtasks.exe' is limited as in it doesn't support
-        # task-time with granularity in seconds. So, falling back
-        # to PowerShell as the last resort.
-        #
+        removed = 0
+        userjs_removed = 0
+        userjs_path = profile_dir + "/user.js"
+        userjs_total = util.count_lines(userjs_path)
 
-        # out_str = ""
-        # task_time = time.strftime("%H:%M",
-        #                           time.localtime(
-        #                               time.time() + 60))
-        #
-        # out_str = subprocess.check_output(
-        #     ["schtasks.exe",
-        #      "/Create",
-        #      "/F",
-        #      "/SC",
-        #      "ONCE",
-        #      "/TN",
-        #      "tridactyl",
-        #      "/TR",
-        #      "calc",
-        #      "/IT",
-        #      "/ST",
-        #      task_time],
-        #     shell=True)
-        # }}}
+        for pref in prefs_to_remove:
+            removed = util.remove_firefox_pref(userjs_path, pref)
+            userjs_removed = userjs_removed + removed
 
-        ff_lock_name = "parent.lock"
+        removed = 0
+        prefsjs_removed = 0
+        prefsjs_path = profile_dir + "/prefs.js"
+        prefsjs_total = util.count_lines(prefsjs_path)
 
-        ff_bin_name = browser_cmd
-        ff_bin_path = '"%s"' % shutil.which(ff_bin_name)
+        for pref in prefs_to_remove:
+            removed = util.remove_firefox_pref(prefsjs_path, pref)
+            prefsjs_removed = prefsjs_removed + removed
 
-        ff_bin_dir = '"%s"' % str(
-            pathlib.WindowsPath(shutil.which(ff_bin_name)).parent
-        )
+        # Removing preferences from 'prefs.js' here turns out to be
+        # insuffcient. Firefox tends to write them back immediately
+        # before quitting from internal memory. The only way around
+        # this so far seems to be the following:
+        #
+        #   1. Manually reset the target preference(s) from
+        #      about:config
+        #
+        #   2. Reset 'prefs.js' _after_ Firefox quit, and _before_
+        #      Firefox restarted
+        #
+        # We are resorting to [2] below for the sake of automation,
+        # and hence introducing the pre-restart hooks below. The
+        # hook functionality is likely to be useful for clean-up
+        # activities, adding/removing preferences etc. later on as
+        # well.
```

Comment:
I believe we shouldn't edit prefs.js at all since user.js overrides prefs.js and the MDN recommends against [against doing it](https://developer.mozilla.org/en-US/docs/Mozilla/Preferences/A_brief_guide_to_Mozilla_preferences#Preferences_Saving).

Reply:
> I believe we shouldn't edit prefs.js at all since user.js overrides prefs.js and the MDN recommends against against doing it.

AFAIK, then the only other option is to manually edit `about:config`. I'd be happy to consider other alternatives to this if you can point me to any. Here's my [original comment][0] around this.

I don't have Mozilla's blessing, but in my opinion `prefs.js` _is_ okay to edit when Firefox is **_not_** running (editing works the same way as editing `user.js`), which is what these patches achieve.

Thank you.

[0]: https://github.com/gsbabil/tridactyl/blob/gsbabil/add-fixamo-on-off-support/native/native_main.py#L787-L802

Reply:
I don't understand why you think we need to edit `pref.js`. If "general.Smoothscroll" is set to "true" in `pref.js`, we can just set it to false in `user.js` and the setting will be updated next time the user restarts firefox. Am I missing something?